src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -348,104 +348,100 @@ private static String dumpSharedArchive(boolean isStatic, String fileName) throw
-            Objects.requireNonNull(currentSuperClass);
-
-            if (name.equals(currentClassName)) {
-                // Note: the following call will call back to <code>this.findClass(name)</code> to
-                // resolve the super types of the named class.
-                return super.findClass(name);
+        private final HashMap<String, Source> sources = new HashMap<>();
+
+        private Source resolveSource(String path) throws IOException {
+            Source source = sources.get(path);
+            if (source != null) {
+                return source;
             }
-            if (name.equals(currentSuperClass.getName())) {
-                return currentSuperClass;
+
+            final var file = new File(path);
+            if (!file.exists()) {
+                throw new IOException("No such file: " + path);
             }
-            if (currentInterfaces != null) {
-                for (Class<?> c : currentInterfaces) {
-                    if (name.equals(c.getName())) {
-                        return c;
-                    }
-                }
+            if (file.isFile()) {
+                source = new JarSource(file);
+            } else if (file.isDirectory()) {
+                source = new DirSource(file);
+            } else {
+                throw new IOException("Not a normal file: " + path);
             }
+            sources.put(path, source);
+
+            return source;
+        }
+
+        /**
+         * Load the class of the given <code>name</code> from the given <code>source</code>.
+         * <p>
+         * All unregistered super classes and interfaces of the named class must already be loaded
+         * by this class loader instance.
+         *
+         * @param name the name of the class to be loaded.
+         * @param source path to a directory or a JAR file from which the named class should be
+         *               loaded.
+         */
+        private Class<?> load(String name, String source) throws IOException {
+            final Source resolvedSource = resolveSource(source);
+            final byte[] bytes = resolvedSource.readClassFile(name);
+            // While executing this VM may invoke loadClass() to load supers and interfaces of
+            // this unregistered class. This should only happen for registered supers and
+            // interfaces because all unregistered ones should have already been loaded by this
+            // class loader. Thus it is safe to delegate their loading to system class loader
+            // (our parent) - this is what the default implementation of loadClass() will do.
+            return defineClass(name, bytes, 0, bytes.length);
+        }
 
+        @Override
+        protected Class<?> findClass(String name) throws ClassNotFoundException {
+            // Unregistered classes should be found in load(...), registered classes should be
+            // handeled by parent loaders
```

Comment:
Hmm, I wonder how the reference to java.lang.Object is resolved in this case. When `CustomLoadee` is parsed, the ClassFileParser needs to resolve is super class, which should result into a call to `UnregisteredClassLoader::findClass()`.

```
            "java/lang/Object id: 0",
            "CustomLoadee id: 1 super: 0 source: .",
```

Reply:
`UnregisteredClassLoader::loadClass` will be called, i.e. the default implementation from `UnregisteredClassLoader::loadClass`, which will first try to delegate parent. The delegation chain will reach the boor loader which will find `java.lang.Object` and thus `UnregisteredClassLoader::findClass()` won't be called.

Reply:
I think the findClass() method can be deleted as it's identical to the one in URLClassLoader.